### ISSUE TITLE
Hit outside of full containment

### ISSUE THREAD

**Issue by FaroutYLq** [FaroutYLq/saltax#67]
**Describe the bug**
ValueError: Hit outside of full containment! This seem to show up in `merged_s2s`.

**To Reproduce**
Salt `022175`.

**Expected behavior**
_A clear and concise description of what you expected to happen._

**Screenshots**
_If applicable, add screenshots to help explain your problem._
```
Traceback (most recent call last):
  File "/home/yuanlq/software/saltax/jobs/job.py", line 38, in <module>
    st.make(strrunid, 'merged_s2s')
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1426, in make
    for _ in self.get_iter(run_ids[0], targets,
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1336, in get_iter
    generator.throw(e)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1307, in get_iter
    for n_chunks, result in enumerate(strax.continuity_check(generator), 1):
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/chunk.py", line 304, in continuity_check
    for chunk in chunk_iter:
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processor.py", line 302, in iter
    raise exc.with_traceback(traceback)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/mailbox.py", line 281, in _send_from
    x = next(iterable)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/overlap_window_plugin.py", line 32, in iter
    yield from super().iter(iters, executor=executor)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/plugin.py", line 499, in iter
    yield self.do_compute(chunk_i=chunk_i, **inputs_merged)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/overlap_window_plugin.py", line 49, in do_compute
    result = super().do_compute(chunk_i=chunk_i, **kwargs)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/plugin.py", line 608, in do_compute
    result = self.compute(**kwargs)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/straxen/plugins/merged_s2s/merged_s2s.py", line 132, in compute
    strax.add_lone_hits(merged_s2s, lh, self.to_pe, n_top_channels=n_top_pmts_if_digitize_top)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processing/peak_merging.py", line 170, in add_lone_hits
    _add_lone_hits(peaks, lone_hits, to_pe, n_top_channels=0)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processing/peak_merging.py", line 189, in _add_lone_hits
    raise ValueError('Hit outside of full containment!')
ValueError: Hit outside of full containment!
Python script failed with exit code 1
```

**Versions**
_Please add the output of:_
```
straxen.print_versions()
Host midway3-0064.rcc.local
 module version                                                                               path  git
 python  3.9.18                          /opt/XENONnT/anaconda/envs/XENONnT_development/bin/python None
  strax   1.5.4   /opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax None
straxen   2.1.5 /opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/straxen None
  cutax  1.15.5                                 /dali/lgrandi/xenonnt/software/cutax/v1.15.5/cutax None
```

All _italic_ comments can be removed from this template.

**Comment by FaroutYLq** [comment]
It is possible that this is a problem related to #72

**Comment by FaroutYLq** [comment]
Let's try if in 2024.02.1 the problem could be solved, with [this bug](https://github.com/XENONnT/straxen/pull/1328) in straxen fix

**Comment by FaroutYLq** [comment]
No the bug is still there, for basically the same error message. One might be related but not exactly the same one is:
```
Traceback (most recent call last):
  File "/home/yuanlq/software/saltax/jobs/job.py", line 72, in <module>
    st.make(strrunid, dt, save=(dt))
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/context.py", line 1576, in make
    for _ in self.get_iter(run_ids[0], targets, save=save, max_workers=max_workers, **kwargs):
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/context.py", line 1472, in get_iter
    generator.throw(e)
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/context.py", line 1439, in get_iter
    for n_chunks, result in enumerate(strax.continuity_check(generator), 1):
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/chunk.py", line 298, in continuity_check
    for chunk in chunk_iter:
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/processor.py", line 312, in iter
    raise exc.with_traceback(traceback)
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/mailbox.py", line 274, in _send_from
    x = next(iterable)
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/plugins/plugin.py", line 505, in iter
    yield from self._iter_compute(chunk_i=chunk_i, **inputs_merged)
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/plugins/plugin.py", line 532, in _iter_compute
    yield self.do_compute(chunk_i=chunk_i, **inputs_merged)
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/plugins/plugin.py", line 619, in do_compute
    result = self.compute(**kwargs)
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/straxen/plugins/peaks/peaks.py", line 53, in compute
    peaks = strax.replace_merged(peaklets[~is_s1], merged_s2s)
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/processing/peak_merging.py", line 119, in replace_merged
    skip_windows = strax.touching_windows(orig, merge)
  File "/opt/XENONnT/anaconda/envs/XENONnT_2024.02.1/lib/python3.9/site-packages/strax/processing/general.py", line 394, in touching_windows
    raise ValueError(f"{names} should have non-negative length!")
ValueError: containers should have non-negative length!
```

**Comment by FaroutYLq** [comment]
See this example. It cannot be reproduced when we are in a notebook and process chunk by chunk.

In the log it looks like:
```
ValueError: Hit outside of full containment!
^MLoading merged_s2s: |██        | 20.30 % [04:56<19:24], #23 (12.82 s). 8.9 MB/s
Traceback (most recent call last):
  File "/home/yuanlq/software/saltax/jobs/job.py", line 87, in <module>
    st.make(strrunid, dt, save=(dt), progress_bar=True)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1576, in make
    for _ in self.get_iter(run_ids[0], targets, save=save, max_workers=max_workers, **kwargs):
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1472, in get_iter
    generator.throw(e)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1439, in get_iter
    for n_chunks, result in enumerate(strax.continuity_check(generator), 1):
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/chunk.py", line 298, in continuity_check
    for chunk in chunk_iter:
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processor.py", line 312, in iter
    raise exc.with_traceback(traceback)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/mailbox.py", line 274, in _send_from
    x = next(iterable)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/overlap_window_plugin.py", line 33, in iter
    yield from super().iter(iters, executor=executor)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/plugin.py", line 505, in iter
    yield from self._iter_compute(chunk_i=chunk_i, **inputs_merged)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/plugin.py", line 532, in _iter_compute
    yield self.do_compute(chunk_i=chunk_i, **inputs_merged)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/overlap_window_plugin.py", line 49, in do_compute
    result = super().do_compute(chunk_i=chunk_i, **kwargs)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/plugin.py", line 619, in do_compute
    result = self.compute(**kwargs)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/straxen/plugins/merged_s2s/merged_s2s.py", line 145, in compute
    strax.add_lone_hits(merged_s2s, lh, self.to_pe, n_top_channels=n_top_pmts_if_digitize_top)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processing/peak_merging.py", line 177, in add_lone_hits
    _add_lone_hits(peaks, lone_hits, to_pe, n_top_channels=0)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processing/peak_merging.py", line 196, in _add_lone_hits
    raise ValueError("Hit outside of full containment!")
ValueError: Hit outside of full containment!
Python script failed with exit code 1
```
While if we try to reproduce it by loading single chunks and compute in a notebook:
```python
import gc

st_salt = saltax.contexts.sxenonnt(
             saltax_mode='salt',
             output_folder='/scratch/midway2/yuanlq/salt/ambe',
             faxconf_version="sr0_v4",
             generator_name='ambe',
             recoil=0,
             mode='all')
st_simu = saltax.contexts.sxenonnt(
             saltax_mode='simu',
             output_folder='/scratch/midway2/yuanlq/salt/ambe',
             faxconf_version="sr0_v4",
             generator_name='ambe',
             recoil=0,
             mode='all')

pgn = st_salt.get_single_plugin("051909", "merged_s2s")
for i in range(5):
    print(i)
    lone_hits = st_salt.get_array("051909", "lone_hits")
    peaklet_classification = st_salt.get_array("051909", "peaklet_classification")
    peaklets = st_salt.get_array("051909", "peaklets", progress_bar=True, _chunk_number=20+i)
    
    if pgn.merge_without_s1:
        peaklets = peaklets[peaklets["type"] != 1]
    
    gap_thresholds = pgn.s2_merge_gap_thresholds
    max_gap = gap_thresholds[0][1]
    max_area = 10 ** gap_thresholds[-1][0]
    
    if "data_top" not in peaklets.dtype.names:
        peaklets_w_field = np.zeros(
            len(peaklets), dtype=strax.peak_dtype(n_channels=pgn.n_tpc_pmts, digitize_top=True)
        )
        strax.copy_to_buffer(peaklets, peaklets_w_field, "_add_data_top_field")
        del peaklets
        peaklets = peaklets_w_field
    
    # Max gap and area should be set by the gap thresholds
    # to avoid contradictions
    start_merge_at, end_merge_at = pgn.get_merge_instructions(
        peaklets["time"],
        strax.endtime(peaklets),
        areas=peaklets["area"],
        types=peaklets["type"],
        gap_thresholds=gap_thresholds,
        max_duration=pgn.s2_merge_max_duration,
        max_gap=max_gap,
        max_area=max_area,
    )
    
    assert "data_top" in peaklets.dtype.names
    
    merged_s2s = strax.merge_peaks(
        peaklets,
        start_merge_at,
        end_merge_at,
        max_buffer=int(pgn.s2_merge_max_duration // np.gcd.reduce(peaklets["dt"])),
    )
    merged_s2s["type"] = 2
    
    # Updated time and length of lone_hits and sort again:
    lh = np.copy(lone_hits)
    del lone_hits
    lh_time_shift = (lh["left"] - lh["left_integration"]) * lh["dt"]
    lh["time"] = lh["time"] - lh_time_shift
    lh["length"] = lh["right_integration"] - lh["left_integration"]
    lh = strax.sort_by_time(lh)
    
    # If sum_waveform_top_array is false, don't digitize the top array
    n_top_pmts_if_digitize_top = pgn.n_top_pmts if "data_top" in pgn.dtype.names else -1

    strax.add_lone_hits(merged_s2s, lh, pgn.to_pe, n_top_channels=n_top_pmts_if_digitize_top)
    gc.collect()
```

**Comment by FaroutYLq** [comment]
In the notebook output, it has no bug. So it might be a chunk edging thing?

```
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
0
Loading lone_hits: 
 100.00 % [00:09<00:00], #16 (0.30 s). 345.9 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklet_classification: 
 100.00 % [00:00<00:00], #4 (0.04 s). 1437.5 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklets: 
 18.53 % [00:00<00:02], #1 (0.67 s). 1713.9 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
1
Loading lone_hits: 
 100.00 % [00:08<00:00], #16 (0.29 s). 398.2 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklet_classification: 
 100.00 % [00:00<00:00], #4 (0.06 s). 1391.0 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklets: 
 19.49 % [00:01<00:05], #1 (1.21 s). 1163.2 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
2
Loading lone_hits: 
 100.00 % [00:10<00:00], #16 (0.31 s). 337.8 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklet_classification: 
 100.00 % [00:00<00:00], #4 (0.12 s). 1525.3 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklets: 
 20.30 % [00:01<00:05], #1 (1.36 s). 897.1 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
3
Loading lone_hits: 
 100.00 % [00:05<00:00], #16 (0.16 s). 641.2 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklet_classification: 
 100.00 % [00:00<00:00], #4 (0.04 s). 1387.5 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklets: 
 21.26 % [00:01<00:05], #1 (1.49 s). 884.2 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
4
Loading lone_hits: 
 100.00 % [00:09<00:00], #16 (0.30 s). 351.2 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklet_classification: 
 100.00 % [00:00<00:00], #4 (0.04 s). 1600.2 MB/s
WARNING:strax:Option diffusion_constant not taken by any registered plugin
WARNING:strax:Option diffusion_constant not taken by any registered plugin
Loading peaklets: 
 22.07 % [00:01<00:04], #1 (1.38 s). 703.3 MB/s
```

**Comment by FaroutYLq** [comment]
Running again in notebook actually worked??
```python
st_salt.make("051909", "merged_s2s", progress_bar=True)
```

**Comment by FaroutYLq** [comment]
However, for another one, it keeps failing at exactly the same place:
```python
st_salt.make("051906", "merged_s2s", progress_bar=True)
```
It gives error in the first chunk:
```
ValueError: Hit outside of full containment!
^MLoading merged_s2s: |          | 0.00 % [00:10<?]
Traceback (most recent call last):
  File "/home/yuanlq/software/saltax/jobs/job.py", line 87, in <module>
    st.make(strrunid, dt, save=(dt), progress_bar=True)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1576, in make
    for _ in self.get_iter(run_ids[0], targets, save=save, max_workers=max_workers, **kwargs):
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1472, in get_iter
    generator.throw(e)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/context.py", line 1439, in get_iter
    for n_chunks, result in enumerate(strax.continuity_check(generator), 1):
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/chunk.py", line 298, in continuity_check
    for chunk in chunk_iter:
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processor.py", line 312, in iter
    raise exc.with_traceback(traceback)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/mailbox.py", line 274, in _send_from
    x = next(iterable)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/overlap_window_plugin.py", line 33, in iter
    yield from super().iter(iters, executor=executor)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/plugin.py", line 505, in iter
    yield from self._iter_compute(chunk_i=chunk_i, **inputs_merged)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/plugin.py", line 532, in _iter_compute
    yield self.do_compute(chunk_i=chunk_i, **inputs_merged)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/overlap_window_plugin.py", line 49, in do_compute
    result = super().do_compute(chunk_i=chunk_i, **kwargs)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/plugins/plugin.py", line 619, in do_compute
    result = self.compute(**kwargs)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/straxen/plugins/merged_s2s/merged_s2s.py", line 145, in compute
    strax.add_lone_hits(merged_s2s, lh, self.to_pe, n_top_channels=n_top_pmts_if_digitize_top)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processing/peak_merging.py", line 177, in add_lone_hits
    _add_lone_hits(peaks, lone_hits, to_pe, n_top_channels=0)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.9/site-packages/strax/processing/peak_merging.py", line 196, in _add_lone_hits
    raise ValueError("Hit outside of full containment!")
ValueError: Hit outside of full containment!
Python script failed with exit code 1
```

**Comment by FaroutYLq** [comment]
And the problem for run `051906` can be reproduced by this
```python
import gc

lone_hits = st_salt.get_array("051906", "lone_hits")
peaklet_classification = st_salt.get_array("051906", "peaklet_classification")
peaklets = st_salt.get_array("051906", "peaklets", progress_bar=True, _chunk_number=0)

i_min = np.where(peaklet_classification['time']==np.min(peaklets['time']))[0][0]
i_max = np.where(peaklet_classification['time']==np.max(peaklets['time']))[0][0]
peaklet_classification = peaklet_classification[i_min:i_max+1]
peaklets['type'] = peaklet_classification['type']

if pgn.merge_without_s1:
    peaklets = peaklets[peaklets["type"] != 1]

gap_thresholds = pgn.s2_merge_gap_thresholds
max_gap = gap_thresholds[0][1]
max_area = 10 ** gap_thresholds[-1][0]

if "data_top" not in peaklets.dtype.names:
    peaklets_w_field = np.zeros(
        len(peaklets), dtype=strax.peak_dtype(n_channels=pgn.n_tpc_pmts, digitize_top=True)
    )
    strax.copy_to_buffer(peaklets, peaklets_w_field, "_add_data_top_field")
    del peaklets
    peaklets = peaklets_w_field

# Max gap and area should be set by the gap thresholds
# to avoid contradictions
start_merge_at, end_merge_at = get_merge_instructions(
    peaklets["time"],
    strax.endtime(peaklets),
    areas=peaklets["area"],
    types=peaklets["type"],
    gap_thresholds=gap_thresholds,
    max_duration=pgn.s2_merge_max_duration,
    max_gap=max_gap,
    max_area=max_area,
)

assert "data_top" in peaklets.dtype.names

merged_s2s = strax.merge_peaks(
    peaklets,
    start_merge_at,
    end_merge_at,
    max_buffer=int(pgn.s2_merge_max_duration // np.gcd.reduce(peaklets["dt"])),
)
merged_s2s["type"] = 2

# Updated time and length of lone_hits and sort again:
lh = np.copy(lone_hits)
del lone_hits
lh_time_shift = (lh["left"] - lh["left_integration"]) * lh["dt"]
lh["time"] = lh["time"] - lh_time_shift
lh["length"] = lh["right_integration"] - lh["left_integration"]
lh = strax.sort_by_time(lh)

# If sum_waveform_top_array is false, don't digitize the top array
n_top_pmts_if_digitize_top = pgn.n_top_pmts if "data_top" in pgn.dtype.names else -1

strax.add_lone_hits(merged_s2s, lh, pgn.to_pe, n_top_channels=n_top_pmts_if_digitize_top)
```

**Comment by FaroutYLq** [comment]
The bug should have been fixed in #92. I think the issue was propagated in a non-debuggable way after numba, because of the channel shift.
